File: linkgame-client/src/main/java/io/github/octcarp/linkgame/client/GameMain.java
package io.github.octcarp.linkgame.client;

import io.github.octcarp.linkgame.client.net.PlayerManager;
import io.github.octcarp.linkgame.client.utils.SceneSwitcher;
import javafx.application.Application;
import javafx.application.Platform;
import javafx.scene.control.Alert;
import javafx.scene.control.ButtonType;
import javafx.scene.image.Image;
import javafx.stage.Stage;

import java.util.Objects;

public class GameMain extends Application {

    public static void main(String[] args) {
        launch(args);
    }

    @Override
    public void init() {
    }

    @Override
    public void start(Stage primaryStage) {
        // init primary stage settings
        initPrimaryStage(primaryStage);

        // show login scene finally
        SceneSwitcher.getInstance().setPrimaryStage(primaryStage);
        SceneSwitcher.getInstance().switchScene("login");  // Default login scene
        primaryStage.show();
    }

    private void initPrimaryStage(Stage primaryStage) {
        // set window title & icon
        primaryStage.getIcons().add(new Image(Objects.requireNonNull(GameMain.class.getResource(
                "/img/main_icon.png")).toExternalForm()));
        primaryStage.setTitle("CS209A Linking Game");

        // handle the close event
        primaryStage.setOnCloseRequest(event -> {
            event.consume();
            Alert alert = new Alert(Alert.AlertType.CONFIRMATION);
            alert.setTitle("Confirm Exit");
            alert.setHeaderText("Are you sure you want to exit?");
            alert.setContentText("Your match step will be lost.");
            alert.showAndWait().ifPresent(response -> {
                if (response == ButtonType.OK) {
                    MyExit();
                }
            });
        });
    }

    private void MyExit() {
        PlayerManager.getInstance().logout();
        Platform.exit();
    }
}

File: game-client/src/main/kotlin/io/github/octcarp/sustech/cs209a/linkgame/client/GameMain.kt
package io.github.octcarp.sustech.cs209a.linkgame.client

import io.github.octcarp.sustech.cs209a.linkgame.client.net.ClientService
import io.github.octcarp.sustech.cs209a.linkgame.client.utils.AlertPopper
import io.github.octcarp.sustech.cs209a.linkgame.client.utils.SceneSwitcher
import io.github.octcarp.sustech.cs209a.linkgame.common.packet.Request
import io.github.octcarp.sustech.cs209a.linkgame.common.packet.RequestType
import javafx.application.Application
import javafx.application.Platform
import javafx.event.EventHandler
import javafx.scene.control.Alert
import javafx.scene.control.ButtonType
import javafx.scene.image.Image
import javafx.stage.Stage
import javafx.stage.WindowEvent
import java.util.*
import java.util.function.Consumer
import kotlin.system.exitProcess

fun main(args: Array<String>) {
    Application.launch(GameMain::class.java, *args)
}

class GameMain : Application() {
    override fun init() {
    }

    override fun start(primaryStage: Stage) {
        // init primary stage settings
        initPrimaryStage(primaryStage)

        // show login scene finally
        SceneSwitcher.setPrimaryStage(primaryStage)
        SceneSwitcher.switchScene("login") // Default login scene
        primaryStage.show()
    }

    private fun initPrimaryStage(primaryStage: Stage) {
        // set window title & icon
        primaryStage.icons.add(
            Image(
                Objects.requireNonNull(
                    GameMain::class.java.getResource(
                        "/img/main_icon.png"
                    )
                ).toExternalForm()
            )
        )
        primaryStage.title = "CS209A Linking Game"

        // handle the close event
        primaryStage.onCloseRequest = EventHandler { event: WindowEvent ->
            event.consume()
            Alert(Alert.AlertType.CONFIRMATION).apply {
                title = "Confirm Exit"
                headerText = "Are you sure you want to exit?"
                contentText = "Your match step will be lost."
            }.showAndWait().ifPresent(Consumer { response: ButtonType ->
                if (response == ButtonType.OK) {
                    myExit()
                }
            })
        }
    }

    private fun myExit() {
        AlertPopper.tryExit = true

        try {
            val request = Request(RequestType.SHUTDOWN)
            ClientService.sendRequest(request)
        } catch (_: Exception) {
            // do nothing
        } finally {
            Platform.exit()
            exitProcess(0)
        }
    }
}